File: r/collatz-conjecture/collatz-conjecture.R
# Name: collatz-conjecture.R
# Purpose: exercism, R track, collatz conjecture: 
#   given a number n, find the number of steps required to reach 1
# Written by Z Knight, 2019.10.16
#
collatz_step_counter <- function(num) {
  if (num < 1) stop("num < 1")
  counter <- 0
  while (num > 1) {
    if (num %% 2 == 0) {
      num <- num / 2
    } else {
      num <- num *3 +1
    }
    counter <- counter +1
  }
  return(counter)
}

collatz_step_counter <- Vectorize(collatz_step_counter)

File: javascript/collatz-conjecture/collatz-conjecture.js
/*
 * Name: collatz-conjecture.js
 * Purpose: exercism, javascript track, collatz-conjecture
 * by Z Knight, 2020.04.04
 */

export const steps = (toCheck) => {
  if (toCheck <= 0) {
    throw new Error('Only positive numbers are allowed');
  }
  if (toCheck == 1) {
    return 0;
  }
  else if (toCheck%2 == 0) {
    return steps(toCheck/2) +1;
  }
  else {
    return steps(3*toCheck+1) +1;
  }
};

